ophyd/userapi/cli_api.py

```diff
@@ -3,123 +3,100 @@
 
 """
 
-# Green = \033[0;32m{}\033[0m
-# Red = \033[0;31m{}\033[0m
-
 from __future__ import print_function
 import time
 import functools
+import sys
+from contextlib import contextmanager, closing
+from StringIO import StringIO
+import collections
+
+from IPython.utils.coloransi import TermColors as tc
 
-from ..controls.positioner import EpicsMotor, Positioner
+from epics import caget, caput
+
+from ..controls.positioner import EpicsMotor, Positioner, PVPositioner
 from ..session import get_session_manager
 
 session_mgr = get_session_manager()
-logger = session_mgr._logger
 
-from epics import caget, caput
-import numpy as np
+try:
+    logbook = session_mgr['olog_client']
+except KeyError:
+    logbook = None
 
 __all__ = ['mov',
            'movr',
            'set_pos',
            'wh_pos',
-           'set_lm'
+           'set_lm',
+           'log_pos',
+           'log_pos_diff'
            ]
 
 # Global Defs of certain strings
 
-STRING_FMT = '^14'
-VALUE_FMT = '^ 14f'
-
-
-def _list_of(value, type_=str):
-    """Return a list of types defined by type_"""
-    if value is None:
-        return None
-    elif isinstance(value, type_):
-        return [value]
-
-    if any([not isinstance(s, type_) for s in value]):
-        raise ValueError("The list is of incorrect type")
-
-    return [s for s in value]
-
+FMT_LEN = 18
+FMT_PREC = 6
 
-def _print_string(val):
-    print('{:{fmt}} '.format(val, fmt=STRING_FMT), end='')
 
+def ensure(*ensure_args):
+    def wrap(f):
+        @functools.wraps(f)
+        def wrapper(*args, **kwargs):
+            # First check if we have an iterable first
+            # on the first arg
+            # if not then make these all lists
+            if len(args) > 0:
+                if not hasattr(args[0], "__iter__"):
+                    args = tuple([[a] for a in args])
+            # Now do type checking ignoring None
+            for arg, t in zip(args, ensure_args):
+                if t is not None:
+                    for x in arg:
+                        if not isinstance(x, t):
+                            raise TypeError("Incorect type in parameter list")
```

Comment:
Dillification for a better tomorrow! (exception string with argument index, what you passed, what's expected would be nice)